[LLDB][Process/Utility] Introduce NativeRegisterContextDBReg class

wangleiat opened the pull request:
Since the setup of debug registers for AArch64 and LoongArch is similar,
we extracted the shared logic from Class:
`NativeRegisterContextDBReg_arm64`
into a new Class:
`NativeRegisterContextDBReg`.
This will simplify the subsequent implementation of hardware breakpoints
and watchpoints on LoongArch.

wangleiat committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.5-bogner

DavidSpickett reviewed (commented):
I started to comment on some style issues here then realised that this code is adapted from `lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.cpp` and I think about 90% of it is the same.

I would prefer that you extract the generic parts of the code into a generic `NativeRegisterContextDBReg` then have Arm64 and LoongArch specialise that.

It's one more level of classes in an already deep hierarchy but in this case I think the overlap is large enough to justify it.

This also solves little things like the name of `g_enable_bit` being too generic for a global in a file that does things other than breakpoints. If it's in a file specific to breakpoints, the name is ok.

`GetWatchpointSize` is an example of something that will need to be per architecture but 
`ClearHardwareBreakpoint` is the same for both so it can go in a base class.

DavidSpickett commented:
And I don't see that LoongArch is doing anything unusual, so I think this is just a refactoring exercise. It is native code so if you don't have an AArch64 machine to build on, I can test the PR for you.

DavidSpickett commented:
We could move all this back into NativeRegisterContextLinux too but A: let's take one step at a time and B: it may be that some architecture does not fit the pattern that AArch64 and LoongArch do.

So in the interests of you getting this done and us having less churn, stick to making a NativeRegisterContextDBReg class.

One of these days maybe we'll hoist it further up.

wangleiat commented:
> We could move all this back into NativeRegisterContextLinux too but A: let's take one step at a time and B: it may be that some architecture does not fit the pattern that AArch64 and LoongArch do.
> 
> So in the interests of you getting this done and us having less churn, stick to making a NativeRegisterContextDBReg class.
> 
> One of these days maybe we'll hoist it further up.

Thank you for your suggestion. I will divide it into two patches:

1. Extract the common parts into the `NativeRegisterContextDBReg` class to simplify `NativeRegisterContextDBReg_arm64`.
2. Based on the above, implement the hardware breakpoint and watchpoint functionality for `LoongArch`.

wangleiat committed:
Introduce NativeRegisterContextDBReg class

Created using spr 1.3.5-bogner

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64.h:
```diff
@@ -234,9 +234,9 @@ class NativeRegisterContextLinux_arm64
   size_t GetFPMRBufferSize() { return sizeof(m_fpmr_reg); }
 
-  llvm::Error ReadHardwareDebugInfo() override;
+  Status ReadHardwareDebugInfo() override;
```
@labath I seem to remember you saying we want to move to or away from Status, if I didn't imagine that, which is it?

wangleiat replied:
Because I noticed that the function calling it returns a `Status` (`virtual Status ClearAllHardwareWatchpoints()` in file lldb/include/lldb/Host/common/NativeRegisterContext.h), and other architectures, such as ARM (not AArch64) and PPC, also return a Status.
Should consistency be maintained here? If `llvm::Error` is required, I will revert it. I also made a mistake—the definition in this file `NativeRegisterContextFreeBSD_arm64.h` hasn't been updated yet. I apologize for my carelessness.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.cpp:
```diff
@@ -274,7 +161,7 @@ uint32_t NativeRegisterContextDBReg_arm64::SetHardwareWatchpoint(
 
   // Setup control value
-  control_value = g_enable_bit | g_pac_bits | GetSizeBits(size);
+  control_value = m_hw_dbg_enable_bit | g_pac_bits | GetSizeBits(size);
```
Perhaps you can also move this sort of thing into a method to be overridden. Then this code can also move to the base class.

wangleiat replied:
I have made updates based on [this](https://github.com/llvm/llvm-project/pull/118770#discussion_r1871170826) feedback.

DavidSpickett reviewed (commented):
I was looking at the Arm (32 bit) code and it's also very similar, so if I find the time I'll migrate this to this base class too.

Edit: Though 32/64 probably makes that annoying, anyway, point is, these patterns are clearly more common than we thought.

See if you can move the use of pac_bits, then I'll test this for you on AArch64.

wangleiat committed:
Extract SetHardware{Breakpoint/Watchpoint}

Created using spr 1.3.5-bogner

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.cpp:
```diff
@@ -272,199 +57,27 @@ uint32_t NativeRegisterContextDBReg_arm64::SetHardwareWatchpoint(
-NativeRegisterContextDBReg_arm64::GetWatchpointSize(uint32_t wp_index) {
-  Log *log = GetLog(LLDBLog::Watchpoints);
-  LLDB_LOG(log, "wp_index: {0}", wp_index);
+NativeRegisterContextDBReg_arm64::MakeControlValue(size_t size,
```
So 1st - yes, this is exactly the sort of thing I meant in my comments.

I think at minimum, the `watch_flags` should be `optional<uint32_t>` but even better would be to have:
MakeBreakControlValue
MakeWatchControlValue

Since I'm pretty sure we always know which one we want to use. This makes it harder to misuse these functions.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.h:
```diff
@@ -9,79 +9,16 @@
 
-  uint32_t m_max_hbp_supported;
-  uint32_t m_max_hwp_supported;
+  bool ValidateWatchpoint(size_t size, lldb::addr_t &addr) override;
```
addr_t is a uint64_t, it doesn't need to be a reference. And it being a non-const ref makes it look like we're going to write back to addr, which we don't.

DavidSpickett replied:
Ok, actually we do :rofl: 

This is unfortunate because "Validate" implies we don't mutate anything.

I would split this into Validate that is purely is it valid yes or no, and AdjustWatchpoint that takes a copy of the address and returns a new value.

DavidSpickett replied:
Also this function has a bug, size is assigned to but it is passed by copy. This was the cause of the test failures.

I suggest you change this to a single AdjustWatchpoint that takes two parameters by reference, and returns true if it was able to adjust it and false otherwise.

DavidSpickett replied:
If we want to prevent this sort of mistake it could be:
```
struct WatchpointDetails {
  size_t size;
  lldb::addr_t addr;
}
std::optional<WatchpointDetails> AdjustWatchpoint(const WatchPointDetails &details);
```
This is at least clearer that:
* This can fail (empty optional)
* New values are returned not assigned by reference

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.h:
```diff
@@ -0,0 +1,90 @@
+  // Default hardware breakpoint length size is 4.
+  // Default hardware breakpoint target address must 4-byte alignment.
+  virtual bool ValidateBreakpoint(size_t size, lldb::addr_t addr) {
+    return (size == 4) && !(addr & 0x3);
```
If these specific checks are common to AArch64 and LoongArch, this is fine. You could remove `virtual` if so.

If this is just the AArch64 checks, I think it might be better to make this an abstract method. Just so that future programmers are forced to confirm what their architecture requires, and they can find existing rules by looking at the arch specific class, rather than having to know that one uses the base implementation.

wangleiat replied:
Yes, it's the same. I will update the comments. Thank you!

DavidSpickett reviewed (commented):
Looking good, I will test this version today.

DavidSpickett commented:
And now I see that x86 has a `NativeRegisterContextDBReg_x86` class.

The way it stores registers is quite different though, so I don't think:
1. We can share that much code without changing how it reads the registers.
2. That it's a problem if we just leave `NativeRegisterContextDBReg_x86` as it is.

So let's ignore that class for now. LoongArch and AArch64 is enough to be dealing with.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.cpp:
```diff
@@ -0,0 +1,400 @@
+  if (watch_flags > 3)
+    return LLDB_INVALID_INDEX32;
+  // Encoding watch_flags
+  auto EncodingWatchFlags = [&]() {
```
This is only used once so it doesn't need to be a lambda.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.cpp:
```diff
@@ -0,0 +1,400 @@
+    return LLDB_INVALID_INDEX32;
+
+  // lldb::eWatchpointKindRead | lldb::eWatchpointKindWrite = 3
+  if (watch_flags > 3)
```
If you're going to do this, just write this in the if or use the switch case that the original code did and the `default` will handle the unknown cases.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.cpp:
```diff
@@ -272,199 +57,27 @@ uint32_t NativeRegisterContextDBReg_arm64::SetHardwareWatchpoint(
+NativeRegisterContextDBReg_arm64::MakeControlValue(size_t size,
+                                                   uint32_t *watch_flags) {
+  // PAC (bits 2:1): 0b10
+  uint32_t pac_bits = (2 << 1);
```
const

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.h:
```diff
@@ -0,0 +1,90 @@
+
+  uint32_t m_max_hbp_supported;
+  uint32_t m_max_hwp_supported;
+  uint32_t m_hw_dbg_enable_bit;
```
This should be const, and initialised in the class constructors.

Or, if AArch64 and LoongArch use the same bit, just init it to 1 here in the base class.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg_arm64.cpp:
```diff
@@ -272,199 +57,27 @@ uint32_t NativeRegisterContextDBReg_arm64::SetHardwareWatchpoint(
+  //      0b00000011 means 2 bytes (addr..addr+1)
+  //      ...
+  //      0b11111111 means 8 bytes (addr..addr+7)
+  auto EncodingSizeBits = [](size_t size) { return ((1 << size) - 1) << 5; };
```
We can avoid the overhead of a lambda (or making the compiler remove the overhead) by doing:
```
size_t encoded_size = ((1 << size) - 1) << 5;
```

DavidSpickett commented:
Something is up with watchpoints:
```
Failed Tests (12):
  lldb-api :: commands/watchpoints/multi_watchpoint_slots/TestWatchpointMultipleSlots.py
  lldb-api :: commands/watchpoints/watchpoint_commands/command/TestWatchpointCommandLLDB.py
  lldb-api :: commands/watchpoints/watchpoint_commands/command/TestWatchpointCommandPython.py
  lldb-api :: commands/watchpoints/watchpoint_commands/condition/TestWatchpointConditionCmd.py
  lldb-api :: commands/watchpoints/watchpoint_size/TestWatchpointSizes.py
  lldb-api :: python_api/watchpoint/condition/TestWatchpointConditionAPI.py
  lldb-api :: tools/lldb-dap/databreakpoint/TestDAP_setDataBreakpoints.py
  lldb-shell :: Subprocess/clone-follow-child-wp.test
  lldb-shell :: Subprocess/clone-follow-parent-wp.test
  lldb-shell :: Subprocess/fork-follow-child-wp.test
  lldb-shell :: Subprocess/fork-follow-parent-wp.test
  lldb-shell :: Subprocess/vfork-follow-parent-wp.test
```
I'm digging into it now.

DavidSpickett commented:
The test failures were caused by:
```
virtual bool ValidateWatchpoint(size_t size, lldb::addr_t &addr) = 0;
```
Which is actually trying to change size for AArch64, but it's passed by copy so it wasn't being updated in the caller. See my comments on those lines.

With that fixed, the test suite passes.

If you address that, I'll test this again. If it's all working then you can address @SixWeining 's style suggestions from the original PR, and I'll have a few of my own.

wangleiat committed:
Address @DavidSpickett's comments

Created using spr 1.3.5-bogner

wangleiat commented:
> The test failures were caused by:
> 
> ```
> virtual bool ValidateWatchpoint(size_t size, lldb::addr_t &addr) = 0;
> ```
> 
> Which is actually trying to change size for AArch64, but it's passed by copy so it wasn't being updated in the caller. See my comments on those lines.
> 
> With that fixed, the test suite passes.
> 
> If you address that, I'll test this again. If it's all working then you can address @SixWeining 's style suggestions from the original PR, and I'll have a few of my own.

Thanks very much for your comments.
If nothing was missed, everything should be fixed.

wangleiat commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_loongarch64.cpp:
```diff
@@ -337,4 +354,471 @@ NativeRegisterContextLinux_loongarch64::GetExpeditedRegisters(
+    return 0;
+  }
+
+  LLDB_LOG(log, "{0}", m_max_hbp_supported);
```
> Could we make the log message more readable? For example `Supported hardware breakpoints: {0}`

`LLDB_LOG` will output the function name, which is sufficient for logging.

wangleiat commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_loongarch64.cpp:
```diff
@@ -337,4 +354,471 @@ NativeRegisterContextLinux_loongarch64::GetExpeditedRegisters(
+
+  return Status();
+}
+bool NativeRegisterContextLinux_loongarch64::BreakpointIsEnabled(
```
This interface has already been moved to `NativeRegisterContextDBReg`. This class is specifically designed for implementing hardware breakpoints and watchpoints, so emphasizing it separately is somewhat redundant.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.cpp:
```diff
@@ -0,0 +1,401 @@
+  if (!ValidateBreakpoint(size, addr))
+    return LLDB_INVALID_INDEX32;
+
+  // Setup control value
```
Comment is now redundant.

DavidSpickett commented on lldb/source/Plugins/Process/Utility/NativeRegisterContextDBReg.cpp:
```diff
@@ -0,0 +1,401 @@
+
+  uint32_t control_value = 0, bp_index = 0;
+
+  // Check hardware breakpoint size and address.
```
This comment is redundant.

wangleiat replied:
Thank you, I will remove the redundant comments.

DavidSpickett reviewed (commented):
This passed all tests. Some remaining cosmetic issues but otherwise looks good.

We could go more into the style issues here but
1. This code was like that before, you just moved it, so it's not actually a regression.
2. I don't want to get into refactoring and break tests again. If we want to do that we can do it in small pieces later.

Going to look again at the follow up PR as well.

wangleiat commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64.cpp:
```diff
@@ -180,7 +180,8 @@ NativeRegisterContextLinux_arm64::NativeRegisterContextLinux_arm64(
                                         register_info_up.release()),
-      NativeRegisterContextLinux(native_thread) {
+      NativeRegisterContextLinux(native_thread),
+      NativeRegisterContextDBReg_arm64(1U) {
```
Thank you, I will add the parameter annotations.

wangleiat commented:
> This passed all tests. Some remaining cosmetic issues but otherwise looks good.
> 
> We could go more into the style issues here but
> 
> 1. This code was like that before, you just moved it, so it's not actually a regression.
> 2. I don't want to get into refactoring and break tests again. If we want to do that we can do it in small pieces later.
> 
> Going to look again at the follow up PR as well.

Do you mean submitting the subsequent changes in a new PR? Since the `llvm::Error -> Status` modification is incomplete, it will cause a compilation error in `NativeRegisterContextFreeBSD_arm64.cpp`. 
Can I continue making modifications in this PR?

wangleiat committed:
Update some comments and restore the return type(Status->llvm::Error)

Created using spr 1.3.5-bogner

wangleiat commented:
Hi @DavidSpickett, the following are the main changes:
1, Modified the commit message.
2, Restored the return value type (Status->llvm::Error). This reduces the number of files modified. 
3, Deleted some redundant comments which you suggested.
4, Removed the parameters of the derived class constructor as you suggested. 

Thank you very much for your review.

DavidSpickett reviewed (approved):
Since you mentioned FreeBSD I tried this on AArch64 FreeBSD and it doesn't cause any test failures (beyond what was already failing on `main`).

This all LGTM.
